segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,182 @@
+            val txn = chunkedSegmentStorage.getMetadataStore().beginTransaction(false, chunkToDelete);
+            val future =
+                    txn.get(chunkToDelete)
+                    .thenComposeAsync( metadata -> {
+                            val chunkMetadata = (ChunkMetadata) metadata;
+                            // Check whether the chunk is marked as inactive
+                            if (!chunkMetadata.isActive()) {
+                                return chunkedSegmentStorage.getChunkStorage().openWrite(chunkToDelete)
+                                        .thenComposeAsync(chunkedSegmentStorage.getChunkStorage()::delete, chunkedSegmentStorage.getExecutor())
+                                        .thenRunAsync(() -> {
+                                            garbageChunks.remove(chunkToDelete);
+                                            txn.delete(chunkToDelete);
+                                            log.debug("{} deleteGarbage - deleted chunk={}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                        }, chunkedSegmentStorage.getExecutor())
+                                        .thenComposeAsync(v -> txn.commit(), chunkedSegmentStorage.getExecutor())
+                                        .exceptionally(e -> {
+                                            val ex = Exceptions.unwrap(e);
+                                            if (ex instanceof ChunkNotFoundException) {
+                                                log.debug("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                            } else {
+                                                log.warn("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                            }
+                                            return null;
+                                        });
+                            } else {
+                                log.info("{} deleteGarbage - Chunk is not marked as garbage{}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                return CompletableFuture.completedFuture(null);
+                            }
+                        })
+                    .whenCompleteAsync((v, ex) -> {
+                        txn.close();
+                    });
+                    futures.add(future);
+        }
+        return Futures.allOf(futures);
+    }
+
+    /**
+     * Marks given chunk as not garbage.
+     *
+     * @param chunkName Name of the chunk.
+     */
+    void removeFromGarbage(String chunkName) {
+        garbageChunks.remove(chunkName);
+    }
+
+    boolean canDelete(long time, long currentTime) {
+        return delayInMillis < (currentTime - time);
+    }
+
+    @Override
+    public void close() throws Exception {
+        closed.set(true);
```

Comment:
In this close, you should probably wait for your `run` future to complete as well. The contract for `close` is that when it exits, any activities started by this instance will have ended.